### PR TITLE
Emit kubernetes events from KEDA

### LINKED ISSUE #530: Emit Kuberentes Events on major KEDA events
KEDA controller should emit Kuberentes Events on major events. This is based on #479 

### Use-Cases
- New Trigger Authentication Created
- Deployment Scaled to Zero
- New Scaled Object Created
- Information on scaling action
- Information about new scaled object referencing existing trigger authentication
- ...

### PR DESCRIPTION
**Pull request by ahmelsayed**
Signed-off-by: Ahmed ElSayed <ahmels@microsoft.com>

This PR adds the following events:

For both `ScaledObjects` and `ScaledJobs`:

- `Ready`
- `CheckFailed`
- `Deleted`
- `ScalersStarted`
- `ScalersRestarted`
- `ScalersStopped`

For `ScaledObjects`:

- `ScaleTargetActivated`
- `ScaleTargetDeactivated`
- `ScaleTargetActivationFailed`
- `ScaleTargetDeactivationFailed`

For `ScaledJobs`:

- `JobsCreated`

If this list looks okay, I'll open a docs PR to list them in there. 

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO)
- [ ] Tests have been added
- [x] A PR is opened to update the documentation on https://github.com/kedacore/keda-docs
- [x] Changelog has been updated

Fixes #530

### CODE REVIEW

File: adapter/main.go
Comment (coderanger) [inline adapter/main.go:7]: Any reason to use this directly instead of the controller-runtime wrappers?

File: pkg/scaling/scale_handler.go
Comment (coderanger) [inline pkg/scaling/scale_handler.go:99]: Would this emit every time the controller pod restarts?

File: controllers/scaledobject_controller.go
Comment (coderanger) [inline controllers/scaledobject_controller.go:171]: This isn't usually the kind of thing you would want in an event since it's not a specific action or event, it's a convergent state.

File: pkg/scaling/scale_handler.go
Comment (ahmelsayed) [reply pkg/scaling/scale_handler.go:100]: Yes. I wasn't sure if there is a best practice somewhere for when to emit events. I tried to add all the ones mentioned in #530 but if there is a best practices for this I'd love to check it

Comment (coderanger) [reply pkg/scaling/scale_handler.go:100]: The basic guideline is it should only be when something actually occurs. So any situation where a reconcile happens and it takes no action, it should produce no events :) Otherwise they can get spammy and overwhelm Etcd.

**Comment by tomkerkhove** [comment]
Before we merge, can you PR a new sub-page to our `Operate` docs please?
https://keda.sh/docs/2.0/operate/

File: controllers/scaledjob_controller.go
Comment (zroubalik) [inline controllers/scaledjob_controller.go:96]: This should be emitted only for the first time.

File: controllers/scaledobject_controller.go
Comment (zroubalik) [reply controllers/scaledobject_controller.go:172]: Agree, we should emit only once, for the first time.

File: pkg/scaling/scale_handler.go
Comment (zroubalik) [reply pkg/scaling/scale_handler.go:100]: But I don't suppose that controller pod restarts happen that often or it is something that we should consider as normal. Not sure what other way we can emit this kind of event?

File: CHANGELOG.md
Comment (zroubalik) [inline CHANGELOG.md:25]:
```suggestion
- Emit Kubernetes Events on KEDA events ([#1523](https://github.com/kedacore/keda/pull/1523))
```
Minor nit to keep the Changelog consistent.

**Review by zroubalik — COMMENTED** [review]
The list of events looks okay!

File: pkg/eventreason/eventreason.go
Comment (tomkerkhove) [inline pkg/eventreason/eventreason.go:19]: This is something I'd like to discuss @ahmelsayed @zroubalik because when you are on the other side of the fence (like I am with https://github.com/tomkerkhove/k8s-event-bridge) it is super important that these are unique and clear.

I'd suggest that we always add a KEDA prefix or suffix, I don't know what you think?

For example `Ready` is very generic and makes sense in this context, but when processing events it's hard to determine what is ready based on the reason. For example what if it's `ScaledObjectIsReady` or `KEDAScaledObjectIsReady`

**Comment by tomkerkhove** [comment]
What about these events:
- Scaler failed (which is not stopped I guess?)
- Scaler unauthorized
- New Trigger Authentication Created
- Deployment Scaled to Zero
- Scaledobject/SCaledJob is linked to trigger authentication

**Comment by ahmelsayed** [comment]
Thanks @coderanger, @zroubalik, @tomkerkhove  for the feedback. 
I made the following changes in https://github.com/kedacore/keda/pull/1523/commits/2872b99f62738433d1b4636d04d819b90676726f


- The `*Ready` events to only happen either on the first time a ScaledObject/ScaledJob is reconciled, or if it's previous Ready status was False/Unknown. 
- Removed `ScalersRestarted` since that would always happen on any ScaledObject/ScaledJob update
- Renamed events to either include `ScaledJob`, `ScaledObject`, or `KEDA` prefix 
- Added a controller for TriggerAuthentication and events for a newly added TriggerAuthentication or deleted one. 
- Opened https://github.com/kedacore/keda-docs/pull/361 

Some remarks:

- `ScalersStarted` event will still fire for all scalers on KEDA restart. 
- @tomkerkhove, Scaler unauthorized will be `KEDAScalerFailed`, but that will also happen for any other errors other than Unauthorized. Scalers themselves don't get passed the event recorder object, so errors from them are opaque to KEDA itself. Initially I wanted to avoid scaler authers having to deal with Kubernetes APIs too much. do you think this is sufficient?
- @tomkerkhove Regarding "Scaledobject/SCaledJob is linked to trigger authentication", I'm not sure how best to do this tbh. This can happen on any ScaledObject/ScaledJob update. I can diff them on every update and emit those events, but currently we don't really store an easily enumerable list of ScaledObjects/ScaledJobs anywhere (each just gets a [context and keeps checking the scalers until the context is canceled](https://github.com/kedacore/keda/blob/8d70252dd42f093844855f9655aeeb71a7403249/pkg/scaling/scale_handler.go#L135-L143)) I'll need to store references to all of them to be able to diff the old vs new, @zroubalik what do you think about that?

File: controllers/triggerauthentication_controller.go
Comment (coderanger) [inline controllers/triggerauthentication_controller.go:47]: This seems like a slightly weird one, but I don't think it will do any harm :)

File: pkg/scaling/scale_handler.go
Comment (coderanger) [inline pkg/scaling/scale_handler.go:99]: I think this would display on every restart of the controller? Can probably drop this and the scalers-stopped events since they don't correspond to actual actions, just internal code state.

**Review by coderanger — COMMENTED** [review]
Overall +1 from me on the API plumbing side.

**Comment by tomkerkhove** [comment]
> * Renamed events to either include `ScaledJob`, `ScaledObject`, or `KEDA` prefix

I've had a look https://github.com/kedacore/keda-docs/pull/361 and it's a bit odd since some have the KEDA prefix and others don't. I know I've requested this, but if others think it's stupid I would remove it or add the prefix to all of them. Thoughts @zroubalik?

Just for context: The reason why I suggested this was for consumers since they typically process the whole event stream and this would make it easier for them to understand where these come from.

> * Opened [kedacore/keda-docs#361](https://github.com/kedacore/keda-docs/pull/361)

Thanks 💘

> * @tomkerkhove, Scaler unauthorized will be `KEDAScalerFailed`, but that will also happen for any other errors other than Unauthorized. Scalers themselves don't get passed the event recorder object, so errors from them are opaque to KEDA itself. Initially I wanted to avoid scaler authers having to deal with Kubernetes APIs too much. do you think this is sufficient?

That's not ideal as you might want to filter out to detect authentication issues, but we can still split them later on if this is too much trouble now. Thoughts @zroubalik?

> * @tomkerkhove Regarding "Scaledobject/SCaledJob is linked to trigger authentication", I'm not sure how best to do this tbh. This can happen on any ScaledObject/ScaledJob update. I can diff them on every update and emit those events, but currently we don't really store an easily enumerable list of ScaledObjects/ScaledJobs anywhere (each just gets a [context and keeps checking the scalers until the context is canceled](https://github.com/kedacore/keda/blob/8d70252dd42f093844855f9655aeeb71a7403249/pkg/scaling/scale_handler.go#L135-L143)) I'll need to store references to all of them to be able to diff the old vs new, @zroubalik what do you think about that?

Let's leave this out then, we can still add it later on if need be?

File: CHANGELOG.md
Comment (zroubalik) [inline CHANGELOG.md:46]: Nit: this should be move to the unreleased

File: controllers/scaledjob_controller.go
Comment (zroubalik) [inline controllers/scaledjob_controller.go:43]: Event recorder for Metrics Adapter is named `keda-metrics-adapter` and this one is named `scale-handler`. For consistency, this one could be maybe named `keda-operator`/ `keda-controller`. WDYT?

Comment (zroubalik) [inline controllers/scaledjob_controller.go:43]: Event recorder for Metrics Adapter is named `keda-metrics-adapter` and this one is named `scale-handler`. For consistency, this one could be maybe named `keda-operator`/ `keda-controller` and sync them with those set in main.go WDYT?

File: CHANGELOG.md
Comment (zroubalik) [inline CHANGELOG.md:46]: Nit: This should be moved to Unreleased section above.

File: controllers/triggerauthentication_controller.go
Comment (zroubalik) [reply controllers/triggerauthentication_controller.go:47]: Yeah, we can go with this for now :)

File: pkg/scaling/scale_handler.go
Comment (zroubalik) [reply pkg/scaling/scale_handler.go:99]: Yeah, but how often do we want to(or would like to see) restart the controller? Is this happening that often? And in fact, with a restart the scalers do start watch, so the event message is correct.

**Review by zroubalik — COMMENTED** [review]
Looking good, I like the renamed event names!

- we should probably add a controller for ClusterTriggerAuthentication and cover this new resource as well.
- ad "Scaledobject/SCaledJob is linked to trigger authentication" discussion - yeah, we would need to track it as you suggest. But I don't think is necessary to add this complexity now. We could add later if there's a need from community. what you say @tomkerkhove?

File: CHANGELOG.md
Comment (zroubalik) [reply CHANGELOG.md:46]: And remove the vim suffix at the end :)

**Comment by zroubalik** [comment]
> > * Renamed events to either include `ScaledJob`, `ScaledObject`, or `KEDA` prefix
> 
> I've had a look [kedacore/keda-docs#361](https://github.com/kedacore/keda-docs/pull/361) and it's a bit odd since some have the KEDA prefix and others don't. I know I've requested this, but if others think it's stupid I would remove it or add the prefix to all of them. Thoughts @zroubalik?
I personally don't have a problem with `ScaledJob`, `ScaledObject`, or `KEDA` prefix.

> 
> Just for context: The reason why I suggested this was for consumers since they typically process the whole event stream and this would make it easier for them to understand where these come from.
> 
> > * Opened [kedacore/keda-docs#361](https://github.com/kedacore/keda-docs/pull/361)
> 
> Thanks 💘
> 
> > * @tomkerkhove, Scaler unauthorized will be `KEDAScalerFailed`, but that will also happen for any other errors other than Unauthorized. Scalers themselves don't get passed the event recorder object, so errors from them are opaque to KEDA itself. Initially I wanted to avoid scaler authers having to deal with Kubernetes APIs too much. do you think this is sufficient?
> 
> That's not ideal as you might want to filter out to detect authentication issues, but we can still split them later on if this is too much trouble now. Thoughts @zroubalik?
+1 split later if needed
> 
> > * @tomkerkhove Regarding "Scaledobject/SCaledJob is linked to trigger authentication", I'm not sure how best to do this tbh. This can happen on any ScaledObject/ScaledJob update. I can diff them on every update and emit those events, but currently we don't really store an easily enumerable list of ScaledObjects/ScaledJobs anywhere (each just gets a [context and keeps checking the scalers until the context is canceled](https://github.com/kedacore/keda/blob/8d70252dd42f093844855f9655aeeb71a7403249/pkg/scaling/scale_handler.go#L135-L143)) I'll need to store references to all of them to be able to diff the old vs new, @zroubalik what do you think about that?
> 
> Let's leave this out then, we can still add it later on if need be?
+1

**Comment by tomkerkhove** [comment]
> I personally don't have a problem with `ScaledJob`, `ScaledObject`, or `KEDA` prefix.

So you're ok with how they are or would you use KEDA prefix for all?

**Comment by zroubalik** [comment]
> > I personally don't have a problem with `ScaledJob`, `ScaledObject`, or `KEDA` prefix.
> 
> So you're ok with how they are or would you use KEDA prefix for all?

I am ok with how they are now.

**Comment by ahmelsayed** [comment]
Regarding the naming, I was initially looking how the default kubernetes events are named, and they were all named as `Created`, `Deleted`, `Killing`, `Pulling`, etc. So I assumed the pattern is to have a verb for the name and drive the meaning from the object the event is on. The current names have the prefix `ScaledObject` or `ScaledJob`, but the ones for scalers are shared for both since scalers are the same regardless of the target.

File: controllers/scaledjob_controller.go
Comment (ahmelsayed) [reply controllers/scaledjob_controller.go:43]: Yeah, I think that makes sense. I'll change it to one recorder with the name `keda-operator`

**Review by tomkerkhove — APPROVED** [review]
Sounds good to me, it's a lot better than the default ones :D

LGTM, let me know when the docs are updated!

**Review by zroubalik — APPROVED** [review]

**Comment by tomkerkhove** [comment]
🚀

File: pkg/scaling/scale_handler.go
Comment (tomkerkhove) [inline pkg/scaling/scale_handler.go:99]: @ahmelsayed Is it possible to add the name of every scaler/trigger here?

File: pkg/scaling/scale_handler.go
Comment (tomkerkhove) [inline pkg/scaling/scale_handler.go:125]: @ahmelsayed Is it possible to add the name of every scaler/trigger here?

**Pull request merged**
